feat: Improve evaluation diagrams

GrayNekoBean opened the pull request:
## Purpose

Improve the diagrams generated during evaluation and red teaming. The following 5 diagrams are currently plotted:
- `evaluation-stat-boxplot`: Box plot diagrams that demonstrate the data distribution of all numeric statistical data other than rating scores
- `evaluation-gpt-boxplot`: A box plot diagram that demonstrates the data distribution of rating scores of all metrics assessed during GPT evaluation
- `evaluation-gpt-radar`: A radar diagram that illustrates the average rating scores of all metrics assessed during GPT evaluation
- `evaluation-results`: Bar charts that demonstrate the summarised results (pass count, pass rate and average rating score) of GPT evaluation
- `red_teaming_results`: A radar chart that summarises the results of the red teaming scorers

## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

- [ ] Yes
- [x] No

## Does this require changes to learn.microsoft.com docs?

This repository is referenced by [this tutorial](https://learn.microsoft.com/azure/developer/python/get-started-app-chat-template)
which includes deployment, settings and usage instructions. If text or screenshot need to change in the tutorial,
check the box below and notify the tutorial author. A Microsoft employee can do this for you if you're an external contributor.

- [ ] Yes
- [x] No

## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Code quality checklist

See [CONTRIBUTING.md](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/CONTRIBUTING.md#submit-pr) for more details.

- [x] The current tests all pass (`python -m pytest`).
- [ ] I added tests that prove my fix is effective or that my feature works
- [ ] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [ ] I ran `python -m mypy` to check for type errors
- [x] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

GrayNekoBean committed:
aggregated the bar charts and added boxplot diagrams

binxuan39 committed:
Add radar chart for evaluation result

binxuan39 committed:
fix: Edit stat boxplot data

binxuan39 committed:
fix: remove log info

GrayNekoBean committed:
minor improvements on diagrams

- Display short metric name for the radar graph
- Made bar chart colourful
- Made y_limit an input parameter for plotting functions

phoevos commented on evaluation/evaluate.py:
I'd suggest that we move all plotting-specific utils to a separate module to potentially reuse them between evaluation approaches!

phoevos replied:
With that in mind, we should remove any "evaluation-specific" logic, e.g. default paths referring to the evaluation.

GrayNekoBean replied:
Yes I was also considering about that, I can do that.

phoevos commented on evaluation/utils.py:
Same comment as above, I'd suggest that we move all plotting-specific utils to a separate module to potentially reuse them between evaluation approaches!

phoevos commented on evaluation/evaluate.py:
```diff
@@ -194,81 +196,224 @@ def run_evaluation_from_config(working_dir: Path, config: dict, num_questions: i
+        "pass_count": "Pass Count",
+        "pass_rate": "Pass Rate",
+        "mean_rating": "Average Rating",
+        "latency": "Latency (sec)",
+        "f1_score": "F1 Score",
+        "answer_length": "Answer Length (words)",
+        "gpt_groundedness": "GPT Groundedness Rating",
+        "gpt_relevance": "GPT Relevance Rating",
+        "gpt_coherence": "GPT Coherence Rating",
+        "gpt_similarity": "GPT Similarity Rating",
+        "gpt_fluency": "GPT Fluency Rating",
+    }
+
+    short_metric_name = {
+        "pass_count": "Pass Count",
+        "pass_rate": "Pass Rate",
+        "mean_rating": "Avg. Rating",
+        "latency": "Latency",
+        "f1_score": "F1 Score",
+        "answer_length": "Length",
+        "gpt_groundedness": "Groundness",
+        "gpt_relevance": "Relevance",
+        "gpt_coherence": "Coherence",
+        "gpt_similarity": "Similarity",
+        "gpt_fluency": "Fluency",
+    }
+
+    metric_value_labels = {
+        "pass_count": "Number of Questions",
+        "pass_rate": "Percentage",
+        "mean_rating": "Rating Score",
+        "latency": "Seconds",
+        "f1_score": "F1 Score",
+        "answer_length": "Words",
+        "gpt_groundedness": "Rating Score",
+        "gpt_relevance": "Rating Score",
+        "gpt_coherence": "Rating Score",
+        "gpt_similarity": "Rating Score",
+        "gpt_fluency": "Rating Score",
+    }
```
Does it make sense to hardcode stuff like that here? What if we introduce new metrics?
I think it'd be better if we introduced the display/short names and labels as class attributes of our metrics (both custom and built-in ones).

GrayNekoBean replied:
surely using a constant dict to state their value isn't the best approach I can tell, and apparently making them class attributes is a better way. I was just thinking is that really worth it to change them, would anyone apart from us really going to add more metrics? (btw "pass_count", "pass_rate" and "mean_rating" still need to be separately handled)

phoevos replied:
> would anyone apart from us really going to add more metrics?

@GrayNekoBean haha that's highly unlikely for this project, but still, maintaining these directly at their source is way more scalable. I would also argue that the short name is probably redundant.

phoevos commented on evaluation/utils.py:
```diff
@@ -20,3 +28,76 @@ def save_jsonl(data: list[dict], path: Path):
+
+def radar_factory(num_vars):
+    """
+    Create a radar chart with `num_vars` Axes.
```
My maths years are well behind me, I'm gonna need some explanation here, too much talk of polar transforms and whatnot :rofl: Why is this needed?

phoevos replied:
Edit: this is coming straight from the `matplotlib` docs, which is perfectly fine, but a comment with a reference is sorely needed: https://matplotlib.org/stable/gallery/specialty_plots/radar_chart.html

phoevos commented on evaluation/evaluate.py:
```diff
@@ -194,81 +196,224 @@ def run_evaluation_from_config(working_dir: Path, config: dict, num_questions: i
+    output_path: Path = Path("evaluation_results.png"),
+):
+    """Plot bar charts for the provided data."""
+    assert layout[0] * layout[1] == len(data), "Number of data points must match the layout"
```
We don't typically raise `AssertionErrors` outside of tests when we can have much richer errors. I'd probably switch to sth like this:
```suggestion
    if layout[0] * layout[1] != len(data):
        raise ValueError("Number of data points must match the layout")
```

GrayNekoBean replied:
sure that makes sense, I'll change them in my next commit along with other fixes.

phoevos replied:
Note that suggestions like that can be committed directly from the GH UI!

binxuan39 replied:
bug: layout is fixed to (1, 3), if we only select 2 of the numeric da, it will raise the Error. mark here, will take a look how to fix this later

phoevos commented on evaluation/evaluate.py:
```diff
@@ -194,81 +196,224 @@ def run_evaluation_from_config(working_dir: Path, config: dict, num_questions: i
+    layout: Tuple[int, int], data: List[List[float]], titles: List[str], y_labels: List[str], output_path: Path
+):
+    """Plot box charts for the provided data."""
+    assert layout[0] * layout[1] == len(data), "Number of data points must match the layout"
```
Same note as above, we don't typically raise `AssertionErrors` outside of tests when we can have much richer errors. I'd probably switch to sth like this:
```suggestion
    if layout[0] * layout[1] != len(data):
        raise ValueError("Number of data points must match the layout")
```

GrayNekoBean replied:
same above

phoevos commented on evaluation/evaluate.py:
```diff
@@ -194,81 +196,224 @@ def run_evaluation_from_config(working_dir: Path, config: dict, num_questions: i
+    plt.close(fig)
+
+
+def plot_box_charts(
```
Why do we need both this and `plot_box_chart`? Can't we abstract the box chart creation in a single function and control any sort of special behaviours through params?

phoevos replied:
Because it looks like this one just executes a more elaborate version of `plot_box_chart`.

GrayNekoBean replied:
Well, we *could* use `plot_box_charts` to plot a 1x1 layout with only one group of data. But it makes function call ugly and the parameter inputs less readable. I still insist on making them two distinct functions: one `plot_box_charts` that plots multiple box plot graphs for one set of data each, and another one `plot_box_chart` that plots one box plot graph for multiple sets of data.

phoevos replied:
@GrayNekoBean I wouldn't mind it too much if `plot_box_charts` at least called the other one. This way it's just weird given that they practically have the same name. If one of them does sth super particular, perhaps the name should portray that. It currently reads "call me to plot any a basic box chart with the input data" and "call me to plot multiple basic box charts with the input data".

phoevos replied:
Once again, note that these will be extended very shortly to allow for comparing LLMs, so they should be written in a flexible way.

phoevos reviewed (changes requested):
Thanks, looks good!

I've left some comments that are mostly related to reusability; note that we're likely to reuse these plotting utils in different places (e.g. red-teaming, LLM comparison), so they shouldn't be tied to the `evaluate` command.

In terms of code-style I might have to recheck the docstrings, but given that the tests are passing we should be ok for the most part.

Apart from that, after we merge the CI PR, remember to include here changes to the workflow so that the new plots are posted as artifacts.

binxuan39 commented on evaluation/evaluate.py:
```diff
@@ -194,81 +196,224 @@ def run_evaluation_from_config(working_dir: Path, config: dict, num_questions: i
 
-    plt.savefig(results_dir / "eval.png")
-    plt.close(fig)
+    plot_box_charts(
+        (1, 3),
+        data[len(gpt_metric_list) :],
+        titles[len(gpt_metric_list) :],
+        y_labels[len(gpt_metric_list) :],
+        results_dir / "evaluation_stat_boxplot.png",
+    )
```
Bug: the layout is fixed to (1,3), if we only select other numbers of numeric metrics, it will cause an Error. Mark here, will take a look at how to fix this later

GrayNekoBean committed:
multiple code reorganize and fixes

- migrated all graph-specific code to a new module `diagram_gen`
- Set display name etc. as the class attribute of metrics
- auto calculate layout based on number of datasets passed in
- separated summarizing function and plotting function
- minor fixes

GrayNekoBean committed:
format code with black

phoevos commented on evaluation/evaluate_metrics/base_metric.py:
```diff
@@ -10,6 +10,11 @@
 class BaseMetric(ABC):
     METRIC_NAME = "name_of_metric"
+    
+    # Constants for diagram generation
+    DISPLAY_TITLE = "Title of Metric Diagram"
+    SHORT_NAME = "Short Name"
+    Y_AXIS_LABEL = "Y Axis Label"
```
I'd suggest that we don't make these specific to plotting (especially given that they're part of the Metric classes). We can rename them to something like:

```python
    DISPLAY_NAME = "Base Evaluation Metric Rating"
    SHORT_NAME = "Base Metric"
    NOTE = "Rating Score"
```

GrayNekoBean replied:
Hmmm makes sense, then can we merge it once they are fixed?

phoevos commented on evaluation/evaluate_metrics/builtin_metrics.py:
Same notes wrt the the class attributes.

phoevos commented on evaluation/evaluate_metrics/code_metrics.py:
Same notes wrt the the class attributes.

phoevos reviewed (changes requested):
Haven't finished with the review, but I would start with this change.

GrayNekoBean committed:
changed class attribute naming

GrayNekoBean commented:
How could the CI just fail on a file I didn't even change??

phoevos reviewed (changes requested):
We should be saving the generated plots in a vector graphics format (e.g. as `.pdf`) so that we can zoom in as much as we want. This is crucial for when they end up in the evaluation report.

Apart from that, the CI workflow should be updated with the paths of the new artifacts (please also sync the branch with `main`).

On top of the above, please extend the red teaming side to generate a `scores.pdf` at the end with a radar chart as we discussed.

Other than that, looks good, please integrate the above and we can merge (I'll still change the package name from `diagram_gen` to sth like `plotting` or `plot_utils`, and a few docstrings that don't match the Google style docstring convention)!

GrayNekoBean committed:
export pdf format diagram by default

GrayNekoBean committed:
Sync with remote-tracking branch 'origin/main'

GrayNekoBean committed:
feat: generate radar graph for red-teaming eval

GrayNekoBean committed:
renamed diagram_gen to plotting

GrayNekoBean committed:
Update .gitignore and launch.json for vsc debugging

GrayNekoBean committed:
update python test snapshot

GrayNekoBean commented:
![image](https://github.com/user-attachments/assets/c47a1cee-7938-4afe-907f-d2730b4bfd0b)

I have no idea why would *mypy* suddenly started to report this error on a file that has zero connection to my commits and screwed up the CI, but my instinct told me that I shouldn't fix it here.

phoevos commented on evaluation/evaluate_metrics/base_metric.py:
```diff
@@ -11,6 +11,11 @@
 class BaseMetric(ABC):
     METRIC_NAME = "name_of_metric"
 
+    # Constants for diagram generation
+    DISPLAY_NAME = "Title of Metric Diagram"
+    SHORT_NAME = "Short Name"
+    NOTE = "Y Axis Label"
```
Please take a look again at my comment [here](https://github.com/ucl-contoso-chat/ucl-openai-search/pull/22#discussion_r1705463485). These metrics have nothing to do with plotting, it's not their responsibility, so there shouldn't be any reference to diagram specific stuff in their implementation.
```suggestion
    DISPLAY_NAME = "Base Evaluation Metric Rating"
    SHORT_NAME = "Base Metric"
    NOTE = "Rating Score"
```

GrayNekoBean replied:
I'll fix the rest of the others you just mentioned. Just two comments:

> Finally, the changes committed as part of .gitignore and .vscode/launch.json are irrelevant, they look like they were carried over from the report effort. Similarly, why are the snapshots updated?

That was a mistake, I accidentally committed to this branch. I'll revert that.

> Similarly, why are the snapshots updated?

pytest failed in my local environment because these snapshots were not updated, but now CI can't even reach pytest because it fails on mypy. According to Raimund maybe it has something to do with pip install? I don't know if the same issue would occur on this branch even after I solved the mypy thing.

phoevos reviewed (changes requested):
@GrayNekoBean, I see that the CI is still not updated with the new file paths! On top of that, what you're doing in `save_figures`, checking the file extensions and forcibly changing them to `.pdf` is super over-engineered IMO. You're not creating an API where users might ask for arbitrary file paths and you **need** to ensure that they're saved as PDFs regardless. Nor is saving as PDF itself that critical. It is your code, only used internally in the script, and it's your responsibility to ask for the right file type to be created. Which, as a matter of fact, you're not doing. Most of the plotting calls in `evaluate.py` still ask for PNG files to be created (therefore relying on said functionality in `save_figures` to fix the file extension), please fix that (one of the default params corresponding to a file path also points to a png file). Apart from that, I'd urge you to think about and standardise your function arguments. I see that the output path is an optional argument for some of them, but mandatory for others. Some of these defaults are even wrong in that they specify a bad relative path. Why don't you drop them completely and ask for the caller to specify the file path? The same sort of inconsistency is seen with the type hints for many of the function arguments all over the place. Given that all of the code base, including the evaluation scripts, use type hints, I would expect that you use them in the newly added arguments as well. Finally, the changes committed as part of `.gitignore` and `.vscode/launch.json` are irrelevant, they look like they were carried over from the report effort. Similarly, why are the snapshots updated?

GrayNekoBean committed:
fixed path parameter issues

GrayNekoBean committed:
Revert "Update .gitignore and launch.json for vsc debugging"

This reverts commit bdbefbe4cb4dc067d58f4dc0a807b2956ae65d8e.

GrayNekoBean committed:
Update evaluation/evaluate_metrics/base_metric.py

Co-authored-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos commented on evaluation/plotting.py:
```diff
@@ -0,0 +1,201 @@
+        os.mkdir(output_path.parent)
+
+    plt.savefig(output_path, bbox_inches="tight", format=format)
+    plt.close(fig)
```
Is there a reason why you're specifically passing the figure here? It feels weird given that you're simply relying on the default when going through the rest of the operations (including `savefig` above).

GrayNekoBean committed:
Merge remote-tracking branch 'origin/main' into evaluation-graph

GrayNekoBean committed:
changed save_fig parameters

GrayNekoBean commented:
@phoevos Everything is now fixed, now I wish this could be merged ASAP. The newly deployed service from the main branch is not working and I can't test it right now.

phoevos reviewed (changes requested):
@GrayNekoBean, how is everything fixed exactly? I've asked for the CI workflow to be tweaked to incorporate these changes thrice by now. I want to be able to run it to inspect the generated graphs.

Apart from that, I still see random arguments without type hints, as well as the issues with the radar chart code I raised during the initial review (I never resolved those comments wrt the different docstring styles used and the lack of reference to the doc you copy-pasted it from).

Why do you need this merged? You can deploy and test from anywhere you want, it doesn't need to merged to main. How is this blocking you?

GrayNekoBean commented:
> I've asked for the CI workflow to be tweaked to incorporate these changes thrice by now. I want to be able to run it to inspect the generated graphs.

I'll make sure to fix that.

> Apart from that, I still see random arguments without type hints, as well as the issues with the radar chart code I raised during the initial review (I never resolved those comments wrt the different docstring styles used and the lack of reference to the doc you copy-pasted it from).

I've checked every file I changed for this PR, and the only places with no type hints are the codes in `radar_factory` function. I'm not sure where Binxuan copied that from but I don't think that's crucial, that function is for internal use only, 

> Why do you need this merged? You can deploy and test from anywhere you want, it doesn't need to merged to main. How is this blocking you?

First, I'm not sure what happened to the backend, but since it's not working right now, I can't test if the evaluation scripts work. It's okay for me to make changes in the CI workflow file. But other than that, I don't want to push new code changes that I can't even test if it works. By now I know everything works fine, so it's better to merge it here.

Second, Rita's latest branch is based on this one, it would be better to merge mine first to avoid potential conflicts as much as possible

GrayNekoBean committed:
update CI workflow path

GrayNekoBean committed:
one more update to CI workflow path

phoevos committed:
fix: Simplify save_figures

Remove unnecessary type checking and casting code for the output path
parameter; this is an internal function and the caller should ensure
that the path is a a proper pathlib.Path (which your type hint says it
is anyway). Simplify the directory creation code following the same
approach found in the rest of the codebase.

Minor:
- Rename function to save_figure
- Drop new line between docstring and code

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Add missing type hint

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Remove empty line after docstring

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Add reference to Matplotlib docs for radar chart

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Renaming box plotting functions

Please have function names make sort of sense.

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos reviewed (approved):
@GrayNekoBean added the missing type hint and reference to the matplotlib docs myself, along with a couple of other changes. I guess it's now good to merge.
